packages/async-flow/src/types.d.ts

```diff
@@ -68,9 +68,11 @@ export type GuestInterface<T> = {
     ? (...args: Parameters<T[K]>) => Promise<R>
     : T[K] extends HostAsyncFuncWrapper
       ? GuestOf<T[K]>
-      : T[K] extends object
-        ? GuestInterface<T[K]>
-        : T[K];
+      : T[K] extends (...args: any[]) => infer R
+        ? (...args: Parameters<T[K]>) => R
```

Comment:
Isn' this just:
```suggestion
      : T[K] extends (...args: any[]) => infer R
        ? T[K]
```
I don't think that's quite right though. What failure did you run into? Please add the cases to `async-flow/test/types.test-d.ts` so we can avoid regressions in these types.

Reply:
Thanks, this is simpler and works. The error I was seeing was: `Type 'GuestInterface<() => HostInterface<ChainAddress>>' has no call signatures.`

I added a type test in [f10bad5](https://github.com/Agoric/agoric-sdk/pull/10045/commits/f10bad5eba688629789f94ad19b9679d91654c92). 

I was expecting to be able to do:
```diff
@@ -1,5 +1,6 @@
 import { expectType } from 'tsd';
 import type { Vow, VowTools } from '@agoric/vow';
+import type { Guarded } from '@endo/exo';
 import type {
   HostOf,
   GuestOf,
@@ -30,6 +31,7 @@ type ExoAPIBase = {
   getValue: () => number;
   setValue: (value: number) => void;
   getCopyData: () => Record<string, number>[];
+  getRemote: () => Guarded<{ doFoo: () => void }>;
 };
 type ExoGuestAPI = ExoAPIBase & {
   getValueAsync: () => Promise<number>;
```

But see:
```
Type 'HostInterface<Guarded<{ doFoo: () => void; }>>' is missing the following properties from type 'RemotableBrand<{}, { doFoo: () => void; } & GetInterfaceGuard<{ doFoo: () => void; }>>': L, R
```

Is `Guarded` the correct way to type a durable remotable?

Reply:
Hmm, `Guarded` should:
```
export type Farable<M extends Methods> = M &
  RemotableBrand<{}, M> &
  RemotableObject;
export type Guarded<M extends Methods> = Farable<M & GetInterfaceGuard<M>>;
```
but RemotableBrand expects these private properties,
```
export declare class RemotableBrand<Local, Remote> {
  /** The local properties of the object. */
  private L: Local;

  /** The type of all the remotely-callable functions. */
  private R: Remote;
}
```

I suppose we need to fix `RemotableBrand` so it doesn't appear to have `L` and `R` as real properties. WDYT @michaelfig ? 

Meanwhile I think it's okay to cast or expect error with an explanation

Reply:
Please don't change `RemotableBrand`... it's carefully tuned to give proper errors when you pass it an object that wasn't created correctly as a Remotable.  Consider it an internal type that makes `E` work with things like Exo-objects, Remotable, Far, and durables, not for use outside that layer.

To me, it seems that `HostInterface` is at fault; it needs to explicitly handle any RemotableBrand that was implied by its template parameter.

Reply:
> Isn't this just:

Not the same, but I'm not sure what was intended.

The more complex version returned the type of just the function itself (such as the type of a simple wrapper function), but the "simpler" version preserved the type of the function including all the subproperties that might be hanging off of it.

Reply:
> Meanwhile I think it's okay to cast or expect error with an explanation

I just left a TODO for now

> but I'm not sure what was intended

The changes were to support: `getChainAddress: () => ChainAddress`